hgapp/crafting/views.py

```diff
@@ -268,9 +273,47 @@ def __get_page_data_and_forms(self, POST=None, early_access=False):
         all_contractor_power_full_ids = set(power_fulls.values_list('id', flat=True))
         additional_artifact_power_full_ids = set()
         for artifact in crafted_artifacts:
+
             current_fulls = set(artifact.power_full_set.values_list('id', flat=True))
             additional_artifact_power_full_ids.update(current_fulls.difference(all_contractor_power_full_ids))
             refundable_fulls = refundable_power_fulls_by_artifact_id[artifact.pk]
+
+            # Get upgradable ones... so, all of powers
+            upgradable_pks = all_contractor_power_full_ids
+            # That are on this artifact
+            upgradable_pks = upgradable_pks.intersection(current_fulls)
+            # That aren't refundable... if they're refundable they'll be auto-refunded if they enter a condition where
+            # they could be otherwise upgraded.
+            # Correction, I need to know if it's being updated in higher level logic.
+            # upgradable_pks = upgradable_pks.difference(refundable_fulls)
```

Comment:
These comments should be deleted if the code they're referencing is also commented.